modules/utxo-lib/src/bitgo/Musig2.ts

```diff
@@ -0,0 +1,154 @@
+      // two possible cases after this:
+      // 1. rootWalletKey derives tapInternalKey - it means non-aggregated tapInternalKey
+      // 2. rootWalletKey doesn't derive tapInternalKey
+      // Neither concerns this function, so skip it.
+      return;
+    }
+
+    const participantsPubkeyKeyValues = psbt.getProprietaryKeyVals(inputIndex, {
+      identifier: PSBT_IN_MUSIG_PARTICIPANT_PUB_KEYS,
+      subtype: PSBT_IN_MUSIG_PARTICIPANT_PUB_KEYS_SUBTYPE,
+    });
+    if (participantsPubkeyKeyValues.length !== 1) {
+      throw new Error(
+        `Invalid number of participant pub keys keydata ${participantsPubkeyKeyValues.length} found instead of 1`
+      );
+    }
+
+    const participantsPubkeyKeyValue = participantsPubkeyKeyValues[0];
+    const participantsPubkeyKeydata = participantsPubkeyKeyValue.key.keydata;
+    if (participantsPubkeyKeydata.length !== 64) {
+      throw new Error(`Invalid keydata size ${participantsPubkeyKeydata.length} for participant pub keys`);
+    }
+    if (!participantsPubkeyKeydata.includes(input.tapInternalKey, 32)) {
+      throw new Error(`No tapInternalKey match with participant pub keys keydata`);
+    }
+    if (participantsPubkeyKeyValue.value.length !== 64) {
+      throw new Error(`Invalid valuedata size ${participantsPubkeyKeyValue.value.length} for participant pub keys`);
+    }
+    const tapOutputKey = participantsPubkeyKeydata.subarray(32);
+
+    const participantPubKeys = [0, 1].map((i) => {
+      return participantsPubkeyKeyValue.value.subarray(i * 32, i * 32 + 32);
+    });
+    if (new Set(participantPubKeys).size !== participantPubKeys.length) {
+      throw new Error(`Duplicate participant pub keys found`);
+    }
+    const myDerivations = participantPubKeys
+      .map((pubKey) => {
+        const tapBip32PubKeys = input.tapBip32Derivation!.filter((bipDv) => {
+          return bipDv.pubkey.equals(pubKey);
+        });
+        if (tapBip32PubKeys.length === 1) {
+          return tapBip32PubKeys[0];
+        }
+        throw new Error(
+          `Participant pub key should match to 1 tapBip32Derivations, but found ${tapBip32PubKeys.length}`
+        );
+      })
+      .filter((bipDv) => {
+        return bipDv.masterFingerprint.equals(rootWalletKey.fingerprint);
+      });
+
+    if (myDerivations.length === 0) {
+      throw new Error('Need one tapBip32Derivation masterFingerprint to match the rootWalletKey fingerprint');
+    }
+
+    const myDerivation = myDerivations.filter((bipDv) => {
+      const node = rootWalletKey.derivePath(bipDv.path);
+      return bipDv.pubkey.equals(toXOnlyPublicKey(node.publicKey));
+    });
+
+    if (myDerivation.length !== 1) {
+      throw new Error('participant pub key should match one tapBip32Derivation');
+    }
+
+    const participantPubkey = myDerivation[0].pubkey;
+
+    const { hash } = psbt.getTxHashForTaprootInput(inputIndex);
+
+    const nonceGenArgs = {
+      sessionId,
+      publicKey: toPlainPublicKey(myDerivation[0].pubkey),
```

Comment:
If the signer derives an x-only key here, it's a bug (which will be detected when we check that the derived key is in the list of participant keys above) so no need to coerce here.

Reply:
Done.